### PR TITLE
S28-2877/mk playback endpoint

### PR DESCRIPTION
**Pull request by LaurenceRawlings**
### Change description

- endpoint for to request playback of a MK asset
    - `/media-service/vod?recodringId=<id>&userId=<id>

> [!IMPORTANT]
> This PR requires manual testing by QA. Please notify the QA team @hmcts/pre-rec-evidence-qa.

**QA instructions**

- use the swagger ui to send a request for an asset in the dev subscription
- use test [demo MK Player](https://app.io.mediakind.com/PRE-MEDIAKIND-DEV/refPlayer) to test the playback

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/preapi/media/MediaKind.java [hunk 1]
```diff
@@ -58,8 +62,62 @@ public MediaKind(
     }
 
     @Override
-    public String playAsset(String assetId) {
-        throw new UnsupportedOperationException();
+    public PlaybackDTO playAsset(String assetName, String userId) {
+        if (getAsset(assetName) == null) {
+            throw new NotFoundException("Asset: " + assetName);
+        }
```
Comment (jasonpaige) [inline src/main/java/uk/gov/hmcts/reform/preapi/media/MediaKind.java:68]: The `getAsset` method takes care of throwing the exception so just let it bubble up to the global exception handler.
```suggestion
        getAsset(assetName);
```

File: src/main/java/uk/gov/hmcts/reform/preapi/media/MediaKind.java [hunk 2]
```diff
@@ -58,8 +62,62 @@ public MediaKind(
...
+                .assetName(assetName)
+                .streamingPolicyName("Predefined_ClearStreamingOnly")
+                .endTime(Instant.now().plusSeconds(3600).toString())
+                .build();
+            try {
+                mediaKindClient.putStreamingLocator(userId, MkCreateStreamingLocator.builder()
+                    .properties(properties).build());
+            } catch (FeignException e) {
+                throw new MediaKindException();
+            }
```
Comment (jasonpaige) [inline src/main/java/uk/gov/hmcts/reform/preapi/media/MediaKind.java:89]: This code is hiding what is happening in the FeignException. We won't get to know if it's a 500,400,404 etc. In this case, as the business logic doesn't depend on what the exception is, you can let the `FeignException` bubble up as there is an entry in the `GlobalExceptionHandler` for it.
```suggestion
            mediaKindClient.putStreamingLocator(userId, MkCreateStreamingLocator.builder()
                .properties(properties).build());
```

File: src/main/java/uk/gov/hmcts/reform/preapi/media/MediaKind.java [hunk 3]
```diff
@@ -58,8 +62,62 @@ public MediaKind(
...
+
+        // TODO check streaming endpoint running + start if not
+        String hostName;
+        try {
+            hostName = mediaKindClient.getStreamingEndpointByName("default").getProperties().getHostName();
+        } catch (FeignException.NotFound e) {
+            throw new NotFoundException("Streaming Endpoint: default");
+        } catch (FeignException e) {
+            throw new MediaKindException();
+        }
```
Comment (jasonpaige) [inline src/main/java/uk/gov/hmcts/reform/preapi/media/MediaKind.java:100]: Let everything but the 404 bubble up
```suggestion
        }
```

File: src/main/java/uk/gov/hmcts/reform/preapi/media/MediaKind.java [hunk 4]
```diff
@@ -58,8 +62,62 @@ public MediaKind(
...
+
+        MkStreamingLocatorPaths paths;
+        try {
+            paths = mediaKindClient.getStreamingLocatorPaths(userId);
+        } catch (FeignException.NotFound e) {
+            throw new NotFoundException("Streaming Locator: " + userId);
+        } catch (FeignException e) {
+            throw new MediaKindException();
+        }
```
Comment (jasonpaige) [inline src/main/java/uk/gov/hmcts/reform/preapi/media/MediaKind.java:109]: Let everything but the 404 bubble up
```suggestion
        }
```

File: src/main/java/uk/gov/hmcts/reform/preapi/media/MediaKind.java [hunk 5]
```diff
@@ -58,8 +62,62 @@ public MediaKind(
...
+
+        var dash = paths.getStreamingPaths().stream().filter(p -> p.getStreamingProtocol().equals("Dash"))
+            .findFirst().orElse(null);
+        var hls = paths.getStreamingPaths().stream().filter(p -> p.getStreamingProtocol().equals("Hls"))
+            .findFirst().orElse(null);
+
```
Comment (jasonpaige) [inline src/main/java/uk/gov/hmcts/reform/preapi/media/MediaKind.java:115]: do you want to throw some sort of exception here if there is no hls and no dash url? is it ok as long as 1 is not null?

File: src/main/java/uk/gov/hmcts/reform/preapi/controllers/MediaServiceController.java [hunk 6]
```diff
@@ -88,6 +90,19 @@ public ResponseEntity<LiveEventDTO> getLiveEvents(@PathVariable String liveEvent
...
+    @PreAuthorize("hasAnyRole('ROLE_SUPER_USER', 'ROLE_LEVEL_1', 'ROLE_LEVEL_2', 'ROLE_LEVEL_3', 'ROLE_LEVEL_4')")
+    public ResponseEntity<PlaybackDTO> getVod(@RequestParam UUID recordingId, @RequestParam String userId) {
+        // todo: dont rely on naming convention, link asset name in db
+        var mediaService = mediaServiceBroker.getEnabledMediaService();
+        var assetName = recordingId.toString().replace("-", "") + "_output";
+        var data = mediaService.playAsset(assetName, userId);
+        if (data == null) {
+            throw new NotFoundException("Asset: " + assetName);
+        }
+        return ResponseEntity.ok(data);
```
Comment (jasonpaige) [inline src/main/java/uk/gov/hmcts/reform/preapi/controllers/MediaServiceController.java:103]: Probably neater to let the service layer throw more specific exceptions and never return null.
```suggestion
        return ResponseEntity.ok(mediaService.playAsset(assetName, userId));
```

**Review by jasonpaige — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/reform/preapi/media/MediaKind.java (hunk 5 above)
Comment (LaurenceRawlings) [reply src/main/java/uk/gov/hmcts/reform/preapi/media/MediaKind.java:105]: Is a not found exception okay or should I make a custom one specifically for this case?

**Pull request closed without merging**